#34 Local state responses and evaluation handling

CarlosNZ opened the pull request:
Addresses #34 and #18.

This PR adds an additional hook to `ApplicationPageWrapper` (renamed from "ApplicationPage") called `useGetResponsesAndElementState`. This creates three state objects at the PageWrapper level:
- `responsesByCode` - all current user responses in the form `{ <questionCode> : <responseValue>... }` )
- `responsesFullByCode` - same as previous but contains the full response object ( `{ text, optionIndex, reference }` )
- `elementsState` - contains evaluated values for `isVisible`, `isEditable`, and `isRequired` for each element, indexed by `elementCode`. This object is recomputed whenever `responsesByCode` changes.

Note, I have not yet implemented a mutation to actually change the responseValues when the user fills in the form, will be done in a separate PR.

In order to see "Visibility condition" in action, change one of the `visibility_condition` fields in `template_element` to `{"value": false}` (or another query that evaluates to false), and notice that question no longer appears in the form.

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+
+## Question/Element types
+
+**Note**: all parameter fields can also have a dynamic query object instead of a primitive. The `evaluateExpression` function will return literal strings (or numbers, booleans) as is. The types described for the parameters below are the type that is expected to be _returned_ from a query expression.
```
Maybe add a link to the description of evaluateExpression (on back-end) here: https://github.com/openmsupply/application-manager-server/wiki/Query-Syntax

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+
+**Note**: all parameter fields can also have a dynamic query object instead of a primitive. The `evaluateExpression` function will return literal strings (or numbers, booleans) as is. The types described for the parameters below are the type that is expected to be _returned_ from a query expression.
+
+### Short Text Input _(not yet implemented -- highest priority)_
```
This one is implemented - maybe not entirely but there is already one in the master branch, and it is showing on the User Registration. I'll create issues for the next ones.

nmadruga replied:
I've added a issue to finish this one considering anything that could be missing: #87

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+
+#### Input parameters (in the `parameters` JSON)
+
+- **label**: `string` -- Text that shows in the HTML "label" attribute of the form element (suggestion: make this interpret Markdown)
```
> **label** - [...] suggestion: make this interpret Markdown

I have added a issue regarding this suggestion #86

CarlosNZ replied:
Thanks.

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+---
+
+### Text Information _(not yet implemented)_
+
```
I think this one is already implemented - just missing the `SummaryView` I'll add an issue about that.

nmadruga replied:
#88

CarlosNZ replied:
Yeah, although it's not implemented "completely" yet, in that it's not handling changes etc.

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+  optionIndex: <integer> (index from the options array)
+  text: <string> (actual text from options array)
+  reference?: To-do: some way to link the response to an entity in database (e.g. an organisation)
+  -- @nmadruga maybe you have an idea of what we'd need here
```
I think to be able to have this as part of the response the **Input** would also need to receive that information... In my thinking the options would receive not only a string per row but what some reference to what it is linked to. But maybe using a different element plugin - e.g. Radio Buttons Link ?

CarlosNZ replied:
Let's discuss soon, but will probably be important when we do #90, etc.

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+---
+
+### Radio Buttons _(not yet implemented)_
+
```
#89

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+---
+
+### Drop-down Selector _(not yet implemented)_
+
```
#90

nmadruga commented on documentation/internal/Element-Type-Specs.md:
```diff
@@ -0,0 +1,129 @@
+---
+
+### Page Break _(not yet implemented)_
+
```
#91

nmadruga reviewed (commented):
Overall comment - Why the file `Overview-of-Navigation-&-URL-structure.md` was renamed to copy?

nmadruga commented on src/containers/Application/ElementsArea.tsx:
```diff
@@ -1,9 +1,10 @@
 import { Button, Container, Grid, Header, Label, Segment } from 'semantic-ui-react'
 import { ApplicationViewWrapper } from '../../elementPlugins'
-import useLoadElements from '../../utils/hooks/useLoadElements'
+import useLoadElements from '../../utils/hooks/useLoadElementsOfSection'
```
```suggestion
import useLoadElementsOfSection from '../../utils/hooks/useLoadElementsOfSection'
```

And rename where is in use. It makes it simpler to use the same name as the file.

CarlosNZ replied:
Yes, good catch. This is what happens when you use the "Rename Symbol" feature in VSCode. Annoying.

nmadruga commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -17,14 +18,27 @@ const ApplicationPage: React.FC = () => {
+    serialNumber: serialNumber as string,
+  })
+
+  console.log('Responses BY Code', responsesByCode)
+  console.log('Responses Full by Code', responsesFullByCode)
+
+  console.log('Element State', elementsState)
```
Please remove logs

nmadruga commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -1,14 +1,15 @@
-import React, { useEffect } from 'react'
+import React, { useState, useEffect } from 'react'
```
```suggestion
import React, { useEffect } from 'react'
```

CarlosNZ commented:
> Overall comment - Why the file `Overview-of-Navigation-&-URL-structure.md` was renamed to copy?

🤦‍♂️ My bad -- I duplicated it as a basis for a new doc file, and I guess I modified the original rather than the copy. Good spotting, fixed now.

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+  const [elementsState, setElementsState] = useState({})
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+  const { data, loading: apolloLoading, error: apolloError } = useGetElementsAndResponsesQuery({
+    variables: { serial: serialNumber },
+  })
```
There is a problem with that, I think. For the cached data to be used it would check by id, and since we aren't passing the id here it is just making another call to the server. I think we can make a different issue to solve that, but just flagging this here. That's why I was storing the Application `id` in the context area, and we would need to use the skip to not call this hook until we have the application loaded. Otherwise it's just making 2 calls to the server.

CarlosNZ replied:
Yes, I was hoping to use the id, but I knew it wouldn't have finished loading yet. @andreievg  has a couple of ideas of how to make one hook wait for the preceding one (he wrote a comment on it somewhere, but can't find right now).

We should optimise this once we've got some of the other things solidly in place.

nmadruga replied:
I think this is what you are looking for: https://github.com/openmsupply/application-manager-web-app/compare/example-skip-on-useQuery

By the way - I'm thinking about changing the cache for `Application` to use the serial as the identifier for applications stored in the cache. I've done this already in the branch that I'm currently working - but I'm still seeing calls to the graphql server when this runs so I'll continue to investigate. 🕵️‍♀️

nmadruga replied:
Yes, it's sorted now on the PR to-be-created. I fixed by using the `serial` as the primary key on the GraphQL queries (no changes on the back-end). And also changing the order and fields queried in each query.

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+}
+
+function checkForApplicationErrors(data: GetElementsAndResponsesQuery | undefined) {
+  if (data?.applicationBySerial) {
```
Check if the result `data` isn't undefined and if `data.applicationBySerial` isn't undefined it's better, so you don't need to be checking inside the scope of the useEffect - since that's called after doing this checks

nmadruga replied:
Oh and `data.applicationBySerial.applicationResponses` just in case

nmadruga replied:
Oops needs to also check for correct responses for: `template` and `template?.templateSections`

CarlosNZ replied:
Okay, hope what I've done is okay now:
```
function checkForApplicationErrors(data: GetElementsAndResponsesQuery | undefined) {
  if (
    !data?.applicationBySerial?.applicationResponses?.nodes ||
    !data?.applicationBySerial?.template?.templateSections
  )
    return 'Data undefined'
  if (data?.applicationBySerial?.applicationResponses?.nodes.length === 0)
    return 'No responses found'
  return null
}
```

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+
+    if (apolloError) return
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses?.nodes
```
```suggestion
    const applicationResponses = data?.applicationBySerial?.applicationResponses.nodes as ApplicationResponse[]
```

And add the import from `graphql`.

CarlosNZ replied:
OK

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    if (applicationResponses) {
+      applicationResponses.forEach((response) => {
+        const code = response?.templateElement?.code
+        if (code) {
+          currentResponses[code] = response?.value?.text
+          currentFullResponses[code] = response?.value
+        }
+      })
+    }
```
```suggestion
    applicationResponses.forEach((response) => {
      const code = response.templateElement?.code
      if (code) {
        currentResponses[code] = response.value?.text
        currentFullResponses[code] = response.value
      }
    })
```

Since you already casted to `ApplicationResponse[]` at the top (and the checks will make sure you have received a valid object response).

CarlosNZ replied:
Thanks.

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses?.nodes
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections?.nodes
```
```suggestion
    const templateSections = data?.applicationBySerial?.template?.templateSections
      .nodes as TemplateSection[]
```

CarlosNZ replied:
Done.

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections?.forEach((sectionNode) => {
+      sectionNode?.templateElementsBySectionId?.nodes.forEach((element) => {
+        if (element) templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
```
```suggestion
templateSections.forEach((sectionNode) => {
      const elementsInSection = sectionNode.templateElementsBySectionId?.nodes as TemplateElement[]
      elementsInSection.forEach((element) => {
         templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
```

nmadruga replied:
> No idea why ...[spread] doesn't work here.

I think the problem is that the types aren't the same. The structure of `TemplateElementState` isn't considering `undefined` which is a possible return from the GraphQL results for each `TemplateElement`.

CarlosNZ replied:
This suggestion doesn't work. `sectionNode` is underlined, with the error `Object is possibly 'null'`.  🤷‍♂️

CarlosNZ replied:
Nevermind, it works after I apply the suggestion further up. 👍 thx.

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  async function evaluateSingleElement(element: any): Promise<EvaluatedElement> {
+    const evaluationParameters = {
+      objects: [responsesByCode, responsesFullByCode], // TO-DO: Also send user/org objects etc.
+      // graphQLConnection: TO-DO
+    }
```
This is not in use?

CarlosNZ replied:
What do you mean? What is not in use?

Do you mean the graphQLConnection? No, there is no expressions that use it yet, but we can pass this object in here quite easily when we need it.

nmadruga replied:
Sorry, the `evaluationParameters`.

CarlosNZ replied:
Ah, right, yes -- good catch, they should be passed in to the `evaluateExpression` function.

nmadruga reviewed (changes requested):
I'm still trying to understand how this works : P

Just a heads up on my first run reviewing this. 
I'll probably use this branch as base for the next issue that I'll be working on - Keep the responses in cache until we send to the server in a set.

I like the idea of passing this objects to `ElementsArea` I'm still not sure why we need the `ElementsFullByCode` but, the elementsState is great. I still see a problem with how we are querying things from the server - not using what is in cache, but like I added to a comment I think this can be treated in a separated issue. I'll add a not in `Issues Backlog`.

CarlosNZ commented on src/utils/graphql/queries/getApplication.query.ts:
```diff
@@ -31,8 +31,15 @@ export default gql`
             value
+            id
             templateElement {
               code
```
Done.

nmadruga commented:
I'm noticing another problem. Not sure if it's related to the changes on this PR.
Steps:
- Go to the Register page
- Click on `User Registration`
- Go to the next back
- Click `Go back` on the browser
- Now the page is not reloading anymore.

The error that is showing in the console:
<img width="556" alt="Screen Shot 2020-11-05 at 4 10 28 PM" src="https://user-images.githubusercontent.com/16461988/98193198-78088a80-1f81-11eb-921a-96e87d3a5317.png">

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,137 @@
+}
+
+function checkForApplicationErrors(data: GetElementsAndResponsesQuery | undefined) {
+  if (
+    !data?.applicationBySerial?.applicationResponses?.nodes ||
+    !data?.applicationBySerial?.template?.templateSections
+  )
+    return 'Data undefined'
+  if (data?.applicationBySerial?.applicationResponses?.nodes.length === 0)
+    return 'No responses found'
+  return null
```
```suggestion
if (!data?.applicationBySerial) return 'Data undefined'
  const application = data?.applicationBySerial as Application
  if (!application.applicationResponses || !application.template?.templateSections)
    return 'Application missing parameters'
  if (application.template?.templateSections.nodes.length === 0) return 'No sections found'
  if (application.applicationResponses.nodes.length === 0) return 'No responses found'
  application.template?.templateSections.nodes.forEach((section) => {
    const missingElements = section?.templateElementsBySectionId.nodes.some((element) => !element)
    if (missingElements) return 'Application missing elements'
  })
  return null
```

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,137 @@
+    setResponsesByCode(currentResponses)
+    setResponsesFullByCode(currentFullResponses)
+    setElementsExpressions(templateElements)
+    setLoading(false)
```
I've found what was causing the problem to refresh the page! This `setLoading` should be moved to the next useEffort. I'll add one suggestion next.

CarlosNZ replied:
Great! Thanks for picking up on this. :)

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,137 @@
+    setLoading(false)
+  }, [data, apolloError])
+
+  useEffect(() => {
+    evaluateElementExpressions().then((result) => setElementsState(result))
+  }, [responsesByCode])
```
```suggestion
useEffect(() => {
    if (responsesByCode)
      evaluateElementExpressions().then((result) => {
        setElementsState(result)
        setLoading(false)
      })
  }, [responsesByCode])
```

nmadruga reviewed (changes requested):
Ok. Second run on reviewing this PR.
I think the error checking should include more cases, suggested a replacement. See what do you think. And also another suggestion to fix the problem with refreshing the page.

CarlosNZ commented:
Have made changes to make sure the page loads properly on refresh.

I've also made small change to the response format so it reads and writes `{value: text}` rather than literal value string. Still need to handle the other response fields, but that'll will come later when we get to plugins that actually use them.

For now, you can see the conditional `isVisible` in action by setting one of the fields visibility condition to something like:

```
{
  "children": [
    {
      "children": [
        {
          "value": {
            "property": "Q1"
          }
        }
      ],
      "operator": "objectProperties"
    },
    {
      "value": "Carl"
    }
  ],
  "operator": "="
}
``` 
(recommending compacting that.)
Now that field will only appear once Q1 has been answered with "Carl".

nmadruga commented on src/utils/hooks/useGetResponsesAndElementState.tsx:
```diff
@@ -0,0 +1,144 @@
+    return elementsState
+  }
+
+  async function evaluateSingleElement(element: any): Promise<EvaluatedElement> {
```
```suggestion
  async function evaluateSingleElement(element: TemplateElementState): Promise<EvaluatedElement> {
```

nmadruga replied:
I'm also thinking how we could combine these three types: `TemplateElementState`, `EvaluatedElement` and (a new one I have created to be used as following (for other PRs):
```
interface ElementState {
  id: number
  title: string
  elementTypePluginCode: string
  section: number
  category: TemplateElementCategory
  isRequired: boolean
  isVisible: boolean
  isValid: boolean
  // validationMessage: string | null
  isEditable: boolean
}

interface ApplicationElementState {
  [key: string]: ElementState
}
```

Since these 3 have so much in common it would be good to have some intersection between types. Can you think on something? I hasn't spent to much time on it, just think it was kind of annoying to have to add fields in these 3 while I was extending this to use somewhere else that required more fields...

CarlosNZ replied:
Okay, have done the Typing on the suggestion.

I'm not sure how to combine these, maybe I'll wait to see your PR and then we'll consider all three of them in its own issue?

nmadruga replied:
Yes, all right. I think it would be better to merge this one already. And then we will tidy up these types altogether in the other PR. I was trying to avoid to many changes to the other one, but it will be alright to wait until then to change this.

nmadruga reviewed (commented):
Hey, I have had another go on this one. Hopefully we can finish this one up tomorrow, my last ones are just about the types. If we can merge them somehow. Or have on use to other, or create some in between are that both use... Just because it was annoying to have to enter fields in many `interface` while I was trying to expand this work in another branch.
